Data/ByteString/Short/Internal.hs

```diff
@@ -446,6 +569,851 @@ concat sbss =
+                       in Just (h, t)
+{-# INLINE uncons #-}
+
+-- | /O(1)/ Extract the first element of a ShortByteString, which must be non-empty.
+-- An exception will be thrown in the case of an empty ShortByteString.
+--
+-- @since 0.11.3.0
+head :: HasCallStack => ShortByteString -> Word8
+head = \sbs -> case null sbs of
+  True -> error "empty ShortByteString"
+  False -> indexWord8Array (asBA sbs) 0
+{-# INLINE head #-}
+
+-- | /O(n)/ Return all the elements of a 'ShortByteString' except the last one.
+-- An exception will be thrown in the case of an empty ShortByteString.
+--
+-- Note: copies the entire byte array
+--
+-- @since 0.11.3.0
+init :: HasCallStack => ShortByteString -> ShortByteString
+init = \sbs ->
+  let l = length sbs
+      nl = l - 1
+  in case null sbs of
+      True -> error "empty ShortByteString"
+      False -> create nl $ \mba -> copyByteArray (asBA sbs) 0 mba 0 nl
+{-# INLINE init #-}
+
+-- | /O(n)/ Extract the 'init' and 'last' of a ByteString, returning Nothing
+-- if it is empty.
+--
+-- @since 0.11.3.0
+unsnoc :: ShortByteString -> Maybe (ShortByteString, Word8)
+unsnoc = \sbs ->
+  let l = length sbs
+      nl = l - 1
+  in if | l <= 0 -> Nothing
+        | otherwise -> let l' = indexWord8Array (asBA sbs) (l - 1)
+                           i = create nl $ \mba -> copyByteArray (asBA sbs) 0 mba 0 nl
+                       in Just (i, l')
+{-# INLINE unsnoc #-}
+
+
+-- ---------------------------------------------------------------------
+-- Transformations
+
+-- | /O(n)/ 'map' @f xs@ is the ShortByteString obtained by applying @f@ to each
+-- element of @xs@.
+--
+-- @since 0.11.3.0
+map :: (Word8 -> Word8) -> ShortByteString -> ShortByteString
+map f = \sbs ->
+    let l = length sbs
+        ba = asBA sbs
+    in create l (\mba -> go ba mba 0 l)
+  where
+    go :: BA -> MBA s -> Int -> Int -> ST s ()
+    go !ba !mba !i !l
+      | i >= l = return ()
+      | otherwise = do
+          let w = indexWord8Array ba i
+          writeWord8Array mba i (f w)
+          go ba mba (i+1) l
+
+
+-- | /O(n)/ 'reverse' @xs@ efficiently returns the elements of @xs@ in reverse order.
+--
+-- @since 0.11.3.0
+reverse :: ShortByteString -> ShortByteString
+reverse = \sbs ->
+    let l = length sbs
+        ba = asBA sbs
+#if MIN_VERSION_base(4,12,0) && defined(SAFE_UNALIGNED)
+    in create l (\mba -> go ba mba l)
+  where
+    go :: BA -> MBA s -> Int -> ST s ()
+    go !ba !mba !l = case l `quotRem` 8 of
+      (0, r) -> void $ goWord8Chunk 0 r
+      (q, 0) -> goWord64Chunk 0 0 q
+      (q, r) -> do
+        i' <- goWord8Chunk 0 r
+        goWord64Chunk i' 0 q
+     where
+      goWord64Chunk !off !i !cl
+        | i >= cl = return ()
+        | otherwise = do
+            let w = indexWord64Array ba (off + (i * 8)) -- 64-bit offset
+            writeWord64Array mba (cl - 1 - i) (byteSwap64 w)
+            goWord64Chunk off (i+1) cl
```

Comment:
`off` and `cl` remain the same in all calls of `goWord64Chunk` which suggests that we can optimize recursive calls to pass `i` only.